### PR TITLE
feat: Support config files with viper

### PR DESCRIPTION
**Pull request by f0ssel**
This supports intaking settings for `coder server` from a config file named `server.yaml` in the `--global-config` directory. It changes the way we do flags to be config first, meaning all settings have a config value and all settings can optionally have a flag. 

We use viper to first set default values defined in the schema for each setting. We look for a config file if that exists and load that in, then we bind each setting to a cli flag if it's defined. Viper automatically maps to environment variables. The intake order is the following:
```viper defaults from schema < values in server.yaml < coder env variable < cli flag```

Checkout `cli/deployment/config.go` for a view of the new deployment config schema.

Net negative loc 🔥 

Proposing the following breaking Changes:
- (hidden) Env var `CODER_INMEMORY` -> `CODER_IN_MEMORY_DATABASE` to match parity with config value `in_memory_database`
- (hidden) Env var `CODER_TEMPLATE_AUTOIMPORT` -> `CODER_AUTO_IMPORT_TEMPLATES`

### CODE REVIEW

File: cli/deployment/config.go [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/cli/cliui"
+	"github.com/coder/coder/cli/config"
+	"github.com/coder/coder/codersdk"
+)
+
+func newConfig() codersdk.DeploymentConfig {
```
Comment (kylecarbs) [inline cli/deployment/config.go:22]: How would I structure an object array in this form?

eg.
```
gitauth:
- client_id: ...
- client_secret: ...
```

**Review by kylecarbs — APPROVED** [review]
This is awesome.

Few questions:
1. How do I structure objects like arrays with keys and values?
2. I see we don't generate a default config right now... should we?

**Comment by kylecarbs** [comment]
Oh @f0ssel there's an empty `server.yaml` at the root which I assume is just a mistake.

**Comment by f0ssel** [comment]
@kylecarbs look at how derp server path is configured. Basically the path for the config can use dot notation to make objects, but each field will still be a top level field in the schema. 

I think generating a default is not a bad idea. Where do you think something like that would live in the repo?

**Comment by kylecarbs** [comment]
I think a default can be done later tbh, this seems good to merge as-is so we can unblock.

Just to confirm, can you manually test a few bespoke existing options to make sure that the environment variables are an exact match? I assume we're kinda hoping that Viper doesn't change anything right?

**Comment by kylecarbs** [comment]
@f0ssel I mean a structured array, not an object.

File: codersdk/deploymentconfig.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+type DeploymentConfigField[T Flaggable] struct {
+	Key        string
+	Name       string
+	Usage      string
+	Flag       string
+	Shorthand  string
+	Enterprise bool
+	Hidden     bool
+	Value      T
+}
```
Comment (Emyrk) [inline codersdk/deploymentconfig.go:81]: Don't these need json tags?

**Review by Emyrk — COMMENTED** [review]
This came out nice :+1:

**Comment by f0ssel** [comment]
> Just to confirm, can you manually test a few bespoke existing options to make sure that the environment variables are an exact match? I assume we're kinda hoping that Viper doesn't change anything right?

Yes, I actually do think a few may change - I think this may be ok if we can document it since otherwise we'll have to manually specify overrides which gets annoying and people who know and expect the viper default will be confused. If I document a list of the changes would that work?

**Comment by f0ssel** [comment]
We are gonna merge this without the structured array support to unblock other work and will make updates soon for it.

**Pull request merged**